feat: add 'Show read notifications' setting

afonsojramos opened the pull request:
## Summary
Adds a new setting to show read notifications alongside unread ones. Fixes #708.

## Changes
- Added `showReadNotifications` setting (default: **disabled**)
- When enabled, fetches both read and unread notifications via `all=true` API parameter
- Read notifications display at 50% opacity for visual differentiation
- "Mark as read" button is hidden on already-read notifications
- Tooltip includes rate limiting warning for users with many notifications

## Why "Mark as read" isn't a toggle
GitHub's REST API only supports one-way transitions: `unread → read → done`. There's no endpoint to mark a notification as unread, so we can't implement a toggle. See [GitHub API docs](https://docs.github.com/en/rest/activity/notifications).

## Test plan
- [x] Toggle setting on → read notifications appear with reduced opacity
- [x] Toggle setting off → only unread notifications shown (default behavior)
- [x] "Mark as read" button hidden on read notifications
- [x] Unit tests pass

setchy commented:
We should also change the look and feel for read NotificationRows - ie: opacity difference, plus as Read State as a Filter Section option for Read and Unread. 

Unfortunately the REST API doesn't indicate a difference between read and done notification types

afonsojramos commented:
@setchy, alright, I think I've landed on a good implementation here. 

1. Default is now false (opt-in, preserves existing behavior)
2. "Mark as read" button is hidden on already-read notifications
3. Read notifications already render at 50% opacity
4. Added all=true API param when setting is enabled to actually fetch read notifications
5. Added rate limiting warning in the tooltip

On the read vs done question:
Done notifications are DELETEd from the API entirely - they never appear in responses. So, the unread boolean cleanly distinguishes read from unread. The concern about differentiating "read + done" vs "read + undone" doesn't apply since done items are gone.

On rate limiting:
The tooltip warns users, and it's opt-in. Happy to consider deferring to when the batched API is ready if you think that's better.

What do you think?

afonsojramos commented on src/renderer/utils/notifications/filters/filter.ts:
```diff
@@ -24,6 +24,11 @@ export function filterBaseNotifications(
     let passesFilters = true;
 
+    // Filter out read notifications if showReadNotifications is disabled
+    if (!settings.showReadNotifications && !notification.unread) {
```
I thought about this, but I think it might be better as a follow-up. The filter would only be useful when `showReadNotifications` is enabled (otherwise, read notifications aren't even fetched), which could be confusing. The current implementation is self-contained - the setting controls what's fetched, and opacity shows the difference.

setchy commented on src/renderer/components/notifications/NotificationRow.tsx:
```diff
@@ -148,6 +148,7 @@ export const NotificationRow: FC<NotificationRowProps> = ({
 
           <HoverButton
             action={actionMarkAsRead}
+            enabled={!isNotificationRead}
```
After further testing, i noticed two more items
1 - the latest order now makes the notification row hover group inconsistent with the repository row hover group.  
2 - if all notifications are "read", then we should probably hide the repository-level "mark as read" option

setchy commented on src/renderer/types.ts:
```diff
@@ -109,6 +109,7 @@ export interface NotificationSettingsState {
   showPills: boolean;
   showNumber: boolean;
   participating: boolean;
+  showReadNotifications: boolean;
```
```suggestion
  fetchReadNotifications: boolean,
```

setchy commented on src/renderer/components/settings/NotificationSettings.test.tsx:
```diff
@@ -271,6 +271,22 @@ describe('renderer/components/settings/NotificationSettings.tsx', () => {
     );
   });
 
+  it('should toggle the showReadNotifications checkbox', async () => {
```
```suggestion
  it('should toggle the fetchReadNotifications checkbox', async () => {
```

setchy commented on src/renderer/components/settings/NotificationSettings.test.tsx:
```diff
@@ -271,6 +271,22 @@ describe('renderer/components/settings/NotificationSettings.tsx', () => {
+
+    expect(updateSettingMock).toHaveBeenCalledTimes(1);
+    expect(updateSettingMock).toHaveBeenCalledWith(
+      'showReadNotifications',
```
```suggestion
      'fetchReadNotifications',
```

setchy commented on src/renderer/components/Sidebar.tsx:
```diff
@@ -91,7 +91,7 @@ export const Sidebar: FC = () => {
           aria-label="Notifications"
           data-testid="sidebar-notifications"
-          description={`${unreadNotificationCount} unread notifications ↗`}
+          description={`${unreadNotificationCount} ${settings.showReadNotifications ? 'notifications' : 'unread notifications'} ↗`}
```
```suggestion
          description={`${unreadNotificationCount} ${settings.fetchReadNotifications ? 'notifications' : 'unread notifications'} ↗`}
```

setchy replied:
This should also be using `notificationCount` from useAppContext.

Similarly in context/App.tsx L294 should be 
```
    const trayCount = status === 'error' ? -1 : notificationCount;
```

setchy commented on src/renderer/components/settings/NotificationSettings.tsx:
```diff
@@ -328,6 +328,31 @@ export const NotificationSettings: FC = () => {
         />
 
+        <Checkbox
+          checked={settings.showReadNotifications}
```
```suggestion
          checked={settings.fetchReadNotifications}
```

setchy commented on src/renderer/components/settings/NotificationSettings.tsx:
```diff
@@ -328,6 +328,31 @@ export const NotificationSettings: FC = () => {
+        <Checkbox
+          checked={settings.showReadNotifications}
+          label="Fetch read notifications"
+          name="showReadNotifications"
```
```suggestion
          name="fetchReadNotifications"
```

afonsojramos commented:
> 1 - When marking a read notification as done, the notification does not get removed it from the list notifications response. The [github.com/notifications](https://github.com/notifications) correctly differentiates between read and done, but the list notifications API response doesn't differentiate the two.
2 - when marking a notification as read and settings.fetchReadNotifications is enabled, then the notification row should be set as Opacity.READ and the exit transition animation prevented

Issue number 2 is easily fixable, but issue number one is a doozy... Here are the options I can think of:
1. Accept the limitation - document it for users
2. Also suppress done animation when fetchReadNotifications=true (prevents confusing "disappear then reappear")
3. Track done notifications locally (more complex, scope creep) but likely the best one, long term... What do you think @setchy ?
